javascript/computer_science/project_hash_map.md

```diff
@@ -0,0 +1,53 @@
+### Introduction
+
+You already know the magic behind hash maps, now it's time to write your own implementation!
+
+#### Limitation
+
+  Before we get started, we need to lay down some ground rules. JavaScript's dynamic nature of array allow us to insert and retrieve indexes that is outside our array size range. Example: if we create an array of size `16` to be our buckets size, nothing stopping us from storing items at index `500`. This defeats the purpose we are trying to demonstrate, so we need to put some self restriction to work around this.
+
+  1. Use the following snippet whenever you access a bucket through an index. We want to throw an error if we try to access an out of bound index:
+
+      ```javascript
+      if (index < 0 || index >= buckets.length) {
+        throw new Error("Trying to access index out of bound");
+      }
+      ```
+
+  1. Inside the class DO NOT use the array method `push()` to insert your nodes into the buckets. Always use indexes.
```

Comment:
Won't our learners be implementing linked lists for buckets? So I don't think we have to worry about them using the `push` method.

Reply:
Linked Lists are the items inside the buckets. The buckets themselves needs to be created using an Array and be accessed through indexes only.

Reply:
That leads me to believe your suggestion for this to be changed was because you believed Linked List were the buckets themselves? 
https://github.com/Mclilzee/curriculum/commit/490c30f43b7ca387c33b47eedc8ddd086c22bde8#diff-2ae1edcda41d250feb4bded479b31f662bd5c7231c17553e9e351b9ba9e02b67L137

I will revert it back so that it is more clear that Nodes as linked lists are _inside_ the buckets

Reply:
I'm not married to any semantic here, just stating what is clear to _me_. If you feel saying "linked list is _inside_ a bucket" is clearer, go with that. I wonder if @JoshDevHub has a preference here.

Reply:
I assumed you thought about buckets being implemented in a Linked List and learners have no access to the `push` method. But since we are on the same page, I don't want learners to use `push` to insert the nodes inside the buckets.

I see your point of view more clearly now, where if you have a linked list inside the buckets, then the push method will not help since you need to traverse that list to find a location for a new node.